File: scala/apis/atrium-api-fluent-en_GB-scala/src/main/scala/ch/tutteli/atrium/api/fluent/en_GB/scala/AnyAssertions.scala
package ch.tutteli.atrium.api.fluent.en_GB.scala

import ch.tutteli.atrium.assertions.Assertion
import ch.tutteli.atrium.creating.Expect
import ch.tutteli.atrium.domain.builders.ExpectImpl
import ch.tutteli.atrium.domain.builders.creating.AnyAssertionsBuilder
import kotlin.reflect.KClass

class AnyAssertions[T](expect: Expect[T]) {

  def toBe(expected: T): Expect[T] = addAssertion(_.toBe(expect, expected))

  def notToBe(expected: T): Expect[T] =
    addAssertion(_.notToBe(expect, expected))

  def isSameAs(expected: T): Expect[T] =
    addAssertion(_.isSame(expect, expected))

  def isNotSameAs(expected: T): Expect[T] =
    addAssertion(_.isNotSame(expect, expected))

  def toBeNullIfNullGivenElse(assertionCreatorOrNull: Expect[T] => Unit)(implicit kClass: KClass[T]) =
    addAssertion(_.toBeNullIfNullGivenElse[T](expect, kClass, assertionCreatorOrNull))

  def notToBeNull()(implicit kClass: KClass[T]): Expect[T] = isA[T]()

  def notToBeNull(assertionCreator: Expect[T] => Unit)(implicit kClass: KClass[T]): Expect[T] =
    isA[T](assertionCreator)

  def isA[TSub]()(implicit kClass: KClass[TSub]): Expect[TSub] =
    ExpectImpl.INSTANCE.getAny.isA(expect, kClass).getExpectOfFeature

  def isA[TSub](
      assertionCreator: Expect[TSub] => Unit
  )(implicit kClass: KClass[TSub]): Expect[TSub] =
    ExpectImpl.INSTANCE.getAny.isA(expect, kClass).addToFeature(assertionCreator)

  val and: Expect[T] = expect
  def and(assertionCreator: Expect[T] => Unit): Expect[T] = expect.addAssertionsCreatedBy(assertionCreator)

  private def addAssertion(f: AnyAssertionsBuilder => Assertion): Expect[T] =
    expect.addAssertion(f(ExpectImpl.INSTANCE.getAny))
}

File: misc/deprecated/domain/robstoll-lib/atrium-domain-robstoll-lib-common/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/anyAssertions.kt
@file:JvmMultifileClass
@file:JvmName("AnyAssertionsKt")

package ch.tutteli.atrium.domain.robstoll.lib.creating

import ch.tutteli.atrium.assertions.Assertion
import ch.tutteli.atrium.creating.Expect
import ch.tutteli.atrium.creating.SubjectProvider
import ch.tutteli.atrium.domain.builders.ExpectImpl
import ch.tutteli.atrium.domain.creating.changers.ChangedSubjectPostStep
import ch.tutteli.atrium.reporting.Text
import ch.tutteli.atrium.translations.DescriptionAnyAssertion.*
import kotlin.jvm.JvmMultifileClass
import kotlin.jvm.JvmName
import kotlin.reflect.KClass

fun <T> _toBe(subjectProvider: SubjectProvider<T>, expected: T) =
    ExpectImpl.builder.createDescriptive(subjectProvider, TO_BE, expected) { it == expected }

fun <T> _notToBe(subjectProvider: SubjectProvider<T>, expected: T) =
    ExpectImpl.builder.createDescriptive(subjectProvider, NOT_TO_BE, expected) { it != expected }

fun <T> _isSame(subjectProvider: SubjectProvider<T>, expected: T) =
    ExpectImpl.builder.createDescriptive(subjectProvider, IS_SAME, expected) { it === expected }

fun <T> _isNotSame(subjectProvider: SubjectProvider<T>, expected: T) =
    ExpectImpl.builder.createDescriptive(subjectProvider, IS_NOT_SAME, expected) { it !== expected }


fun <T : Any?> _toBeNull(subjectProvider: SubjectProvider<T>) =
    ExpectImpl.builder.createDescriptive(subjectProvider, TO_BE, Text.NULL) { it == null }

fun <T : Any> _toBeNullIfNullGivenElse(
    expect: Expect<T?>,
    type: KClass<T>,
    assertionCreatorOrNull: (Expect<T>.() -> Unit)?
): Assertion =
    if (assertionCreatorOrNull == null) ExpectImpl.any.toBeNull(expect)
    else notToBeNull(expect, type, assertionCreatorOrNull)

private fun <T : Any> notToBeNull(
    expect: Expect<T?>,
    type: KClass<T>,
    assertionCreator: Expect<T>.() -> Unit
) = ExpectImpl.any.notToBeNull(expect, type).collect(assertionCreator)


fun <T, TSub : Any> _isA(
    expect: Expect<T>,
    subType: KClass<TSub>
): ChangedSubjectPostStep<T, TSub> =
    ExpectImpl.changeSubject(expect).reportBuilder()
        .downCastTo(subType)
        .build()
